### ISSUE TITLE
Bug: [no-unnecessary-condition] using nullish coalescing assignment operator in reduce seems to flag false positive

### ISSUE THREAD

**Issue by treskilion** [typescript-eslint/typescript-eslint#7610]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.2.2&fileType=.ts&code=MYewdgzgLgBBIFsCmBBATmghgTxgXhgG0BYAKBhgG8YBLAEwC4YAiegeTCWYBoYpsADkibN%2BQlMxgBfbmQrV6I9px59BwlmKQBhSTLlVajFvQAqAdxCqtIrROmzyhxSboWrvG5vUAhPY-kjJTcACzQkLk91W3V7KTIAXQBuMjItGABlRCRTdXwYACUkUDQ6AB5oNBowAHNeSuqawgSAPhTSMnhkdCxsADpwugBXYCQyrORcoRaAChmkMCgqpAheGigkLA2kAEp8FqoDUEhYBToooWl89c3Mbfaj8GgYBaXcAlfliEJ6BJgAfn%2BBGa7Qon36AiGEBCMy0OweTnCUCGaDAL0WX3aMioUnhQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK%2BSpPRRE0aB2iRwYAL4gtQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```typescript
const someArray = [
  { id: "idOne", type: "typeA" },
  { id: "idOne", type: "typeC" },
  { id: "idTwo", type: "typeA" },
  { id: "idTwo", type: "typeB" },
  { id: "idThree", type: "typeA" }
];

type SomeType = Record<string, string[]>;

someArray.reduce<SomeType>((entries, iteratee) => {
  const { id, type } = iteratee;

  const entry = entries[id] ??= [];
  entry.push(type);

  return entries;
}, {});
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-unneccessary-condition": "error"
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true,
    "noUncheckedIndexedAccess": true
  }
}
```

### Expected Result

Since entries has a type of `Record<string, string[]>` item fetched via index will give either an item or undefined. 
TypeScript seems to recognize this
![image](https://github.com/typescript-eslint/typescript-eslint/assets/44638557/01cb693b-a360-4ce9-84c8-c482795dbf20)

### Actual Result

@typescript-eslint is under the impression that it will always get a result from accessing the object via index and flags the use of `??=` as an unnecessary condition.

**Comment by bradzacher** [comment]
[Based on your example](https://www.typescriptlang.org/play?strict=true&exactOptionalPropertyTypes=true#code/MYewdgzgLgBBIFsCmBBATmghgTxgXhgG0BYAKBhgG8YBLAEwC4YAiegeTCWYBoYpsADkibN+QlMxgBfbmQrV6I9px59BwlmKQBhSTLlVajFvQAqAdxCqtIrROmzyhxSboWrvG5vUAhPY-kjJTcACzQkLk91W3V7KTIAXQBuMjItGABlRCRTdXwYACUkUDQ6AB5oNBowAHNeSuqawgSAPhTSMnhkdCxsADpwugBXYCQyrORcoRaAChmkMCgqpAheGigkLA2kAEp8FqoDUEhYBToooWl89c3Mbfaj8GgYAA98haWaFcJ6ZIMAen+FAoAD0APypJzHZ4fNC4Aiwr4QH50BIwMFggjNdoUWH9ARDCAhGZaHYPJzhKBDNBgGCIlbtGRUKRkoA) the rule is working as expected - the type of `entries[id]` is never nullish.

**Comment by bradzacher** [comment]
Note that you forgot to turn on `noUncheckedIndexedAccess` in your playground link.

---

Same as #6926, #6632, #6635
Please use the search.
